simulator/authorization_manager.go

```diff
@@ -96,3 +104,89 @@ func (m *AuthorizationManager) RetrieveRolePermissions(req *types.RetrieveRolePe
 		},
 	}
 }
+
+func (m *AuthorizationManager) AddAuthorizationRole(req *types.AddAuthorizationRole) soap.HasFault {
+	body := &methods.AddAuthorizationRoleBody{}
+
+	for _, role := range m.RoleList {
+		if role.Name == req.Name {
+			body.Fault_ = Fault("", &types.AlreadyExists{})
+			return body
+		}
+	}
+
+	m.RoleList = append(m.RoleList, types.AuthorizationRole{
+		Info: &types.Description{
+			Label:   req.Name,
+			Summary: req.Name,
+		},
+		RoleId:    m.nextId,
+		Privilege: updateSystemPrivileges(req.PrivIds),
+		Name:      req.Name,
+		System:    false,
+	})
+
+	m.nextId++
+
+	body.Res = &types.AddAuthorizationRoleResponse{}
+
+	return body
+}
+
+func (m *AuthorizationManager) UpdateAuthorizationRole(req *types.UpdateAuthorizationRole) soap.HasFault {
+	body := &methods.UpdateAuthorizationRoleBody{}
+
+	for _, role := range m.RoleList {
+		if role.Name == req.NewName && role.RoleId != req.RoleId {
+			body.Fault_ = Fault("", &types.AlreadyExists{})
+			return body
+		}
+	}
+
+	for i, role := range m.RoleList {
+		if role.RoleId == req.RoleId {
+			m.RoleList[i].Name = req.NewName
+
+			if req.PrivIds != nil {
+				m.RoleList[i].Privilege = updateSystemPrivileges(req.PrivIds)
+			}
+
+			body.Res = &types.UpdateAuthorizationRoleResponse{}
+			return body
+		}
+	}
+
+	body.Fault_ = Fault("", &types.NotFound{})
+
+	return body
+}
+
+func (m *AuthorizationManager) RemoveAuthorizationRole(req *types.RemoveAuthorizationRole) soap.HasFault {
+	body := &methods.RemoveAuthorizationRoleBody{}
+
+	for i, role := range m.RoleList {
+		if role.RoleId == req.RoleId {
+			m.RoleList = append(m.RoleList[:i], m.RoleList[i+1:]...)
+
+			body.Res = &types.RemoveAuthorizationRoleResponse{}
+			return body
+		}
+	}
+
+	body.Fault_ = Fault("", &types.NotFound{})
+
+	return body
+}
+
+func updateSystemPrivileges(privileges []string) []string {
+OUT:
+	for _, spr := range systemPrivileges {
+		for _, pr := range privileges {
+			if spr == pr {
+				continue OUT
```

Comment:
Curious about the use of a label.  Isn't it the same in this case if you remove the label and just `break` here instead?

Reply:
yes, i can. but I have to use another boolean to judge whether to append it to the list. and this has label. i don't know. I picked this because it has less LOC...

Reply:
Ok fine to leave it as-is, I was just curious (I haven't used labels before either)